ui/apps/platform/src/Containers/ComplianceEnhanced/ScanConfigs/Wizard/ClusterSelection.tsx

```diff
@@ -1,9 +1,162 @@
+        <Link to={clustersBasePath}>
+            <Button variant="primary">Install cluster</Button>
+        </Link>
+    );
+}
+
+function ClusterSelection({ clusters, isFetchingClusters }: ClusterSelectionProps): ReactElement {
+    const { hasReadWriteAccess } = usePermissions();
+    const hasWriteAccessForCluster = hasReadWriteAccess('Cluster');
+    const { setFieldValue, values: formikValues }: FormikContextType<ScanConfigFormValues> =
+        useFormikContext();
+
+    const clusterIsPreSelected = useCallback(
+        (row) => formikValues.clusters.includes(row.id),
+        [formikValues.clusters]
+    );
+
+    const { allRowsSelected, selected, onSelect, onSelectAll } = useTableSelection(
+        clusters,
+        clusterIsPreSelected
+    );
+
+    useEffect(() => {
+        const clusterIds = clusters.map((cluster) => cluster.id);
+        const selectedClusterIds = clusterIds.filter((_, index) => selected[index]);
+        if (!isEqual(selectedClusterIds, formikValues.clusters)) {
+            setFieldValue('clusters', selectedClusterIds);
+        }
+    }, [selected, formikValues.clusters, setFieldValue, clusters]);
+
+    function renderTableContent() {
+        return clusters?.map(({ id, name }, rowIndex) => (
+            <Tr key={id}>
+                <Td
+                    key={id}
+                    select={{
+                        rowIndex,
+                        onSelect,
+                        isSelected: selected[rowIndex],
+                    }}
+                />
+                <Td>{name}</Td>
+            </Tr>
+        ));
+    }
+
+    function renderLoadingContent() {
+        return (
+            <Tr>
+                <Td colSpan={8}>
+                    <Bullseye>
+                        <Spinner isSVG />
+                    </Bullseye>
+                </Td>
+            </Tr>
+        );
+    }
+
+    function renderEmptyContent() {
+        return (
+            <Tr>
+                <Td colSpan={6}>
+                    <Bullseye>
+                        <EmptyStateTemplate
+                            title="No clusters found"
+                            headingLevel="h2"
+                            icon={SearchIcon}
+                        >
+                            {hasWriteAccessForCluster && (
+                                <Flex direction={{ default: 'column' }}>
+                                    <FlexItem>
+                                        <Text>Install a cluster to get started</Text>
+                                    </FlexItem>
+                                    <FlexItem>
+                                        <InstallClustersButton />
+                                    </FlexItem>
+                                </Flex>
+                            )}
+                        </EmptyStateTemplate>
+                    </Bullseye>
+                </Td>
+            </Tr>
+        );
+    }
+
+    function renderTableBodyContent() {
```

Comment:
What are pro and con for `return` statement to have the markup versus call secondary functions?

Reply:
I think modularity is kind of what I was going for. To me this feels easier to read when the component is a little more complex. I personally prefer trying to avoid having too much conditional logic within the markup itself.

Reply:
Aha, you remind me for the future that an application of modularity to a somewhat different situation than this is when a function within a components becomes complicated enough, lift up either to module scope or utils file so that it has explicit arguments instead of some of its inputs implicit from closure scope.